### DISCUSSION TITLE
Modular use of flask-login

### DISCUSSION THREAD

**Discussion by artgoldberg** [maxcountryman/flask-login#739] [Q&A] [answered]
I'm a big proponent of modular code because I think it's easier to design, write, test and maintain.

Therefore, I'd like to make modular use of flask-login. That is, I'd like to create a Python module (file) that defines all the generic functions and decorators described in the [Usage documentation](https://pypi.org/project/Flask-Login/), have the Flask app factory initialize it with the `app`, and then use the decorators, like `login_required` that it defines. 

But I don't see an obvious way to do this because of the interdependencies between the flask-login code and the Flask app factory.

I welcome any advice on how to achieve the modularity I want, or an argument that the modularity I want cannot be reasonably achieved.

Thanks
Arthur

**Comment by artgoldberg** [comment]
Hi Folks

Thanks for your great work.

I'm trying to stimulate a response to my question by adding this comment because I'd love to be able to put (almost all) of my code that uses flask-login in a Python module that's separate from the app factory.

Thanks
Arthur

**Comment by davidism** [comment]
It's the same pattern as any other Flask extension.

```python
login = LoginManager()

def create_app():
    app = Flask(__name__)
    login.init_app(app)
    return app
```

Now you can import `login` anywhere without circular dependencies.

**Reply by artgoldberg** [reply]
Thanks David @davidism . I understand that snippet.

But I'm trying to make the custom login code modular. Using the example in the [Project description](https://pypi.org/project/Flask-Login/), how would I put this code in `custom_login.py`:

```
# Our mock database.
users = {'foo@bar.tld': {'password': 'secret'}}

class User(flask_login.UserMixin):
    pass


@login_manager.user_loader
def user_loader(email):
    if email not in users:
        return

    user = User()
    user.id = email
    return user


@login_manager.request_loader
def request_loader(request):
    email = request.form.get('email')
    if email not in users:
        return

    user = User()
    user.id = email
    return user

@app.route('/login', methods=['GET', 'POST'])
def login():
    if flask.request.method == 'GET':
        return '''
               <form action='login' method='POST'>
                <input type='text' name='email' id='email' placeholder='email'/>
                <input type='password' name='password' id='password' placeholder='password'/>
                <input type='submit' name='submit'/>
               </form>
               '''

    email = flask.request.form['email']
    if email in users and flask.request.form['password'] == users[email]['password']:
        user = User()
        user.id = email
        flask_login.login_user(user)
        return flask.redirect(flask.url_for('protected'))

    return 'Bad login'
```
and then use `@flask_login.login_required` in the app factory?
```
@app.route('/protected')
@flask_login.login_required
def protected():
    return 'Logged in as: ' + flask_login.current_user.id
```

Thanks
Arthur

PS. How do you tell Markdown to highlight Python's syntax?

**Reply by davidism** [reply, marked as answer]
If you have an `app` object, then you're not using the factory pattern. Remove that and use blueprints.

**Reply by artgoldberg** [reply]
Thanks @davidism . I'll do that when I have time.

Arthur
